modules/lib/core/index.d.ts

```diff
@@ -4,11 +4,65 @@ declare global {
     }
 }
 
-export type UserKey = `user:${string}:${string}`;
-export type GroupKey = `group:${string}:${string}`;
-export type RoleKey = `role:${string}`;
+export type UserKey<
+	IdProvider extends string = string,
+	Login extends string = string
+> = `user:${IdProvider}:${Login}`;
 
-export type PrincipalKey = UserKey | GroupKey | RoleKey;
+export type GroupKey<
+	IdProvider extends string = string,
+	Name extends string = string
+> = `group:${IdProvider}:${Name}`;
+
+export type RoleKey<T extends string = string> = `role:${T}`;
```

Comment:
I'm not a fan of these changes to `UserKey`, `GroupKey` and `RoleKey`. I think they become more complex  to read and reason about.

Reply:
In @enonic/js-utils I made seperate Generics.

`export type RoleKeyGeneric<T extends string> = `role:${T}``

Reply:
Perhaps these are even overkill, that literal strings are enough.
Such "extensions" could simply continue to live in @enonic/js-utils